### PR TITLE
coq_8_14: init at 8.14+rc1

### PR DESCRIPTION
**Pull request by Zimmi48**
###### Motivation for this change

Coq 8.14+rc1 was just tagged.

###### Things done

Currently testing this, but I expect this will work given that the Coq derivation has been 8.14-ready for a while now.

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

cc @vbgl

### CODE REVIEW

**Comment by Zimmi48** [comment]
`coq_8_14` does build fine on my machine. I'm currently testing at https://github.com/coq-community/coq-nix-toolbox/actions/runs/1245604615 if there are packages in the already quite large `coqPackages_8_14` set that are not actually buildable.

**Comment by vbgl** [comment]
Build failure (is `which` missing from the `nativeBuildInputs`?):

> dev/tools/make_git_revision.sh: line 3: which: command not found

Also `coqPackages_8_14.addition-chains` fails to evaluate; suggested patch:

~~~
--- a/pkgs/development/coq-modules/addition-chains/default.nix
+++ b/pkgs/development/coq-modules/addition-chains/default.nix
@@ -11,7 +11,7 @@ mkCoqDerivation {
 
   inherit version;
   defaultVersion = with versions; switch coq.coq-version [
-    { case = isGe "8.11"; out = "0.4"; }
+    { case = range "8.11" "8.13"; out = "0.4"; }
   ] null;
 
   propagatedBuildInputs = [ mathcomp-ssreflect mathcomp-algebra paramcoq ];
~~~

File: pkgs/development/coq-modules/bignums/default.nix [hunk 1]
```diff
@@ -8,6 +8,7 @@ with lib; mkCoqDerivation {
   defaultVersion = if versions.isGe "8.5" coq.coq-version
     then "${coq.coq-version}.0" else null;
 
+  release."8.14.0".sha256 = "sha256:0jsgdvj0ddhkls32krprp34r64y1rb5mwxl34fgaxk2k4664yq06";
```
Comment (vbgl) [inline pkgs/development/coq-modules/bignums/default.nix:11]: Is the `sha:` prefix intended?

**Comment by Zimmi48** [comment]
Is the problem really that `which` is missing (I think this is just a warning) or this line: `Error: no rule to make target bin/votour (or missing .PHONY)`? How come it didn't fail when I built it on my side?

File: pkgs/development/coq-modules/bignums/default.nix (hunk 1 above)
Comment (Zimmi48) [reply pkgs/development/coq-modules/bignums/default.nix:11]: Lately, I've stopped using the `nix-prefetch-url` tool to favor the `lib.fakeSha256` technique and copying the output of the expected hash. This is how it gets printed nowadays.

Comment (vbgl) [reply pkgs/development/coq-modules/bignums/default.nix:11]: Good to know. Thanks.

**Comment by vbgl** [comment]
Indeed, the `bin/votour` missing target seems to be the actual error. How does it only happen on CI?

BTW, `Cheerios`, `Verdi`, `CoqEAL` still fail to evaluate.

**Comment by Zimmi48** [comment]
`dev/tools/make_git_revision.sh: line 3: which: command not found` can indeed also been seen in the log of the successful build at: https://github.com/coq-community/coq-nix-toolbox/pull/65/checks?check_run_id=3632607063. For some reason, in this case this is a different derivation that is built. However, on my system, I have checked that the derivation that I've built is the same as the one that has failed in CI here. So, that's strange indeed.

**Comment by Zimmi48** [comment]
I also got this line in the log linked above: `make[1]: 'bin/votour' is up to date.`

**Comment by Zimmi48** [comment]
It's a bit annoying to have to manually tweak the `defaultVersion` computation of packages just because they have dependencies that are marked as broken (even if they might actually work fine with Coq 8.14 if we update the dependency). Would it be possible to automatically mark as broken the reverse dependencies of a broken package?

**Comment by vbgl** [comment]
You may leave out the evaluation failures for now and just be sure that they get fixed before 8.14.0 gets in `nixpkgs`.

**Comment by Zimmi48** [comment]
Yes, but I also want CI on the coq-nix-toolbox repo to pass fully, so I'll fix all the evaluation failures.

**Comment by Zimmi48** [comment]
IIUC the only blocking thing to merge this PR would be to understand the CI failure, but despite the failure being reproducible on ofborg runs, it is truly incomprehensible since it never occurred while building locally or on the coq-nix-toolbox CI (and I have checked that the derivations being built are the same). Are there some known outside parameters that can influence how a derivation is built? I believe that my initial build was sandboxed as it ran on my NixOS machine.

**Comment by SkySkimmer** [comment]
https://github.com/NixOS/nixpkgs/blob/2131f28314d353e326eea14dd8523a4befb34316/pkgs/applications/science/logic/coq/default.nix#L165 bin/votour is not a target since the switch to dune

**Comment by Zimmi48** [comment]
Thanks for your suggestion. I will make it conditional. But I still fail to understand why the same version was building fine elsewhere.

**Comment by Zimmi48** [comment]
Rebuilding `coq_8_14` after this change did not trigger a new build on my machine, which seems to indicate that the derivation which was build was already ignoring this target (or the `buildFlags` entirely).

**Comment by Zimmi48** [comment]
This fixed the issue on Linux but we still have a Darwin-specific issue that appears when unpacking the tarball:

```
checkdir error:  cannot create coq-8.14-rc1/test-suite/misc/deps/+?+?
                 Illegal byte sequence
                 unable to process coq-8.14-rc1/test-suite/misc/deps/+?+?/.
checkdir error:  cannot create coq-8.14-rc1/test-suite/misc/deps/+?+?
                 Illegal byte sequence
                 unable to process coq-8.14-rc1/test-suite/misc/deps/+?+?/+?+?.v.
checkdir error:  cannot create coq-8.14-rc1/test-suite/misc/deps/+?+?
                 Illegal byte sequence
                 unable to process coq-8.14-rc1/test-suite/misc/deps/+?+?/+?+?.v.
do not know how to unpack source archive /private/tmp/nix-build-source.drv-0/V8.14+rc1.zip
```

This is very likely related to the tests using Unicode file names introduced in https://github.com/coq/coq/pull/1041 and probably requires an alternative unpacking strategy. However, this is again something that we should have hit before and I don't understand why we didn't...

**Comment by SkySkimmer** [comment]
Something has been broken with the darwin build for a while, although previously in a different way.
eg https://github.com/NixOS/nixpkgs/runs/1959980556 (coq_8_13: 8.13.0 → 8.13.1) has no log just "This check concluded as stale."
The latest darwin success I found was on https://github.com/NixOS/nixpkgs/pull/70642 (coq_8_10: 8.10+β3 → 8.10.0)
We may note that this is before https://github.com/NixOS/nixpkgs/pull/105877#discussion_r550751319 (the specific sub-discussion may be relevant) which changed how the sources are fetched. This PR has unclear darwin status, there is a green mark for `grahamcofborg-eval-darwin` but it does not link to a check page or log so who knows what it means.

**Comment by Zimmi48** [comment]
This one seems to have worked, though: https://github.com/NixOS/nixpkgs/runs/3666496063

**Comment by SkySkimmer** [comment]
Maybe we should just make a rc2 with the unicode files generated dynamically.

**Comment by vbgl** [comment]
IIRC, dune is not yet able to install Coq correctly: the installation phase has to be fixed.

**Comment by vbgl** [comment]
This might have worked about one year ago:
~~~
  installPhase = ''
    runHook preInstall
    dune install --sections=lib,stublibs --libdir=$OCAMLFIND_DESTDIR $pname
    dune install --sections=lib_root,bin,doc,man --prefix=$out $pname
    ln -s $OCAMLFIND_DESTDIR/coq/plugins $out/lib/coq/
    runHook postInstall
  '';
~~~

**Comment by Zimmi48** [comment]
@vbgl I can blindly apply this suggested `installPhase` for the Coq package, but how do I test that things are fine (and that things aren't already fine with the current version)? All I have been testing so far is whether Coq packages build fine using the Coq Nix Toolbox CI and there have been no issue of that sort.

**Comment by vbgl** [comment]
This of course no longer works…

To test the installation, you can look at the result and check were the various files are put.

You can also run `ocamlfind query coq`.

**Comment by Zimmi48** [comment]
> You can also run `ocamlfind query coq`.

OK, except that this has to be `ocamlfind query coq-core` to be relevant for Coq 8.14 (for instance, so that SerAPI works).

I propose to use the normal installation procedure, then using `mv` or `ln` to fix what is incorrect in the resulting layout. I've done a minimal patch to make `ocamlfind query` work and I tested it by building SerAPI with it. Cf. my last two commits.

**Comment by vbgl** [comment]
Oh, nice. I did not notice that you had fixed it.

**Comment by Zimmi48** [comment]
That's new from yesterday :wink: 

FTR, here is the current structure of `result`:

- `bin/`
- `lib/coq/theories`
- `lib/coq/user-contribs`
- `lib/coq-core/META` (and OCaml libs)
- `lib/coqide/META`
- `lib/coqide-server/META` (and OCaml libs)
- `lib/ocaml/4.10.2/site-lib/coq-core` (symbolic link to `lib/coq-core`)
- `lib/stublibs`
- `nix-support`
- `share/applications`
- `share/coq`
- `share/doc` (with just `LICENSE` and `README` files under `coq-core`, etc.)
- `share/man/man1`
- `share/pixmaps`
- `share/texmf/tex/latex/misc`

Do you see anything to fix? From my side, all I noticed is that there should probably also be a symlink from `lib/ocaml/4.10.2/site-lib/coqide-server` to `lib/coqide-server`.

**Comment by Zimmi48** [comment]
> From my side, all I noticed is that there should probably also be a symlink from `lib/ocaml/4.10.2/site-lib/coqide-server` to `lib/coqide-server`.

Now fixed.

**Comment by Zimmi48** [comment]
> Something has been broken with the darwin build for a while, although previously in a different way.
> eg https://github.com/NixOS/nixpkgs/runs/1959980556 (coq_8_13: 8.13.0 → 8.13.1) has no log just "This check concluded as stale."

It seems that this happens because ofborg will sometimes never run these Darwin tests. And this seems to be happening again now.

**Comment by Zimmi48** [comment]
FTR I did some tests on macOS via the Coq Nix Toolbox CI, and there doesn't seem to be any issue building Coq. Could there be again some discrepancy between how CI behaves when run by ofborg or in the GitHub Actions from the Coq Nix Toolbox? Hard to tell if the CI here doesn't complete.

**Comment by Zimmi48** [comment]
The ofborg CI tests on darwin have finally run, and they have completed with a failure. Meanwhile, the same tests ran in the GitHub Actions CI of the Coq Nix Toolbox and have completed successfully.

Once again, building the same derivation was attempted in both cases:

```
/nix/store/sq9wmdfqyppqbhzfrhqv1sc7i03a8f1q-coq-8.14+rc1.drv
```

The failure in ofborg tests is deterministic, it is always:

```
trying https://github.com/coq/coq/archive/V8.14+rc1.zip
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   117  100   117    0     0    609      0 --:--:-- --:--:-- --:--:--   609
100 8749k    0 8749k    0     0  5208k      0 --:--:--  0:00:01 --:--:-- 9053k
unpacking source archive /private/tmp/nix-build-source.drv-0/V8.14+rc1.zip
checkdir error:  cannot create coq-8.14-rc1/test-suite/misc/deps/+?+?
                 Illegal byte sequence
                 unable to process coq-8.14-rc1/test-suite/misc/deps/+?+?/.
checkdir error:  cannot create coq-8.14-rc1/test-suite/misc/deps/+?+?
                 Illegal byte sequence
                 unable to process coq-8.14-rc1/test-suite/misc/deps/+?+?/+?+?.v.
checkdir error:  cannot create coq-8.14-rc1/test-suite/misc/deps/+?+?
                 Illegal byte sequence
                 unable to process coq-8.14-rc1/test-suite/misc/deps/+?+?/+?+?.v.
do not know how to unpack source archive /private/tmp/nix-build-source.drv-0/V8.14+rc1.zip
builder for '/nix/store/p66rcxr6g0q9sn57spbgmn4qq7vsn95d-source.drv' failed with exit code 1
```

Similarly, the failure we were getting on Linux before fixing the build targets was also deterministic (and was not observed in the GitHub Actions CI of the Coq Nix Toolbox, despite the two buildings the same derivation).

I'm quite annoyed by this discrepancy, which is not supposed to happen with Nix. Could the explanation be that the two use a different version of Nix? The Coq Nix Toolbox CI uses Nix 2.3.15, but I don't know how to tell which version the ofborg CI uses.

**Comment by Zimmi48** [comment]
cc @grahamc in case you are able to help with this (see comment just above)

**Comment by Zimmi48** [comment]
So, it's actually not a Nix version problem. I have been able to reproduce the problem observed here in the GitHub Actions CI with Nix 2.3.15. See https://github.com/Zimmi48/coq-nix-toolbox/runs/3771003867?check_suite_focus=true (not explicitly pinned) and https://github.com/Zimmi48/coq-nix-toolbox/runs/3770951230?check_suite_focus=true (explicitly pinned).
Now, what I don't understand is what is the difference that made `/nix/store/sq9wmdfqyppqbhzfrhqv1sc7i03a8f1q-coq-8.14+rc1.drv` build successfully in https://github.com/coq-community/coq-nix-toolbox/runs/3730072982?check_suite_focus=true but consistently fail in other runs (like the two linked above).

**Comment by Zimmi48** [comment]
>  Now, what I don't understand is what is the difference that made `/nix/store/sq9wmdfqyppqbhzfrhqv1sc7i03a8f1q-coq-8.14+rc1.drv` build successfully in https://github.com/coq-community/coq-nix-toolbox/runs/3730072982?check_suite_focus=true but consistently fail in other runs (like the two linked above).

I'm thinking the explanation might simply be that the archive to download is the same on Linux and Darwin. So it suffices that the archive has already been downloaded on Linux and cached in NixOS official cache (or some Cachix) to make it available to derivations building Coq on Darwin. If that's the case, then it alleviates the problem for users (and explains that it wasn't noticed when running CI with the coq-community Cachix, which already contained the derivation output on Linux).

Does that make sense?

In any case, this issue does not seem to be a regression from 8.14 but rather a problem that was introduced with the switch from `fetchTarball` to `fetchzip` in #105877. We should discuss how to resolve it independently of this PR which has been delayed for too long already.

@vbgl Do you agree?

**Review by vbgl — COMMENTED** [review]
The `multinomials` library at version 1.5.4 is compatible with Coq 8.14.

**Comment by Zimmi48** [comment]
With Cyril, we've discussed that we should report the issues with non-reproducibility of derivation building that we've encountered while preparing this PR (the `bin/votour` issue and the macOS-specific failure to download the source tarball) but I don't know when I'll find sufficient time to sit down and write a proper report for this. Probably not for a while.

**Pull request merged**